### PR TITLE
#3392 save CPP tap changer internal variables in dump

### PR DESCRIPTION
**Pull request by vermeulenthi**
**Checklist before requesting a review**

*use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes*

- [x] unit tests and non-regression tests were added (new model, new feature and bug fix)
- [ ] main documentation was updated (update of input/output file, 3rd party, model, repository organization, solver)
- [ ] example documentations were updated (new example in examples folder)
- [x] the corresponding milestone was added in the ticket and in this PR
- [ ] if this PR modifies the parameters or inputs/outputs of a model/solver: the corresponding xsl was added in util/xsl and platform DB were updated
- [ ] if this PR modifies a dictionary: the corresponding french dictionary was updated

### CODE REVIEW

**Comment by vermeulenthi** [comment]
Open questions: 
- New file "-internalvariables.bin" is created. Another option would have been to recycle the ModelCPP dump/load parameter (less code change)
- All ModelPhaseTapChanger and ModelRatioTapChanger, minus the 'side' string, are dumped (does not include attributes from the parent ModelTapChanger class). Check which would be interesting to dump.
- StringSteam is passed through all functions, to dump only 1 file for the ModelNetwork. Another option would be to fetch data from included model and do all the serialization work in the ModelNetwork class.

File: dynawo/sources/Simulation/DYNSimulation.cpp [hunk 1]
```diff
@@ -1503,6 +1503,7 @@ Simulation::dumpState(const boost::filesystem::path& dumpFile) {
 
   model_->dumpParameters(mapValues);
   model_->dumpVariables(mapValues);
+  model_->dumpInternalVariables(mapValues);
```
Comment (rosiereflo) [inline dynawo/sources/Simulation/DYNSimulation.cpp:1506]: Should be hidden inside the dumpVariables

File: dynawo/sources/Modeler/Common/DYNModel.h [hunk 2]
```diff
@@ -422,6 +422,21 @@ class Model {
    */
   virtual void loadVariables(const std::map< std::string, std::string>& mapVariables) = 0;
 
+  /**
+   * @brief export the internal variables of the model for dump
+   *
+   * @param mapInternalVariables : map associating the file where values should be dumped with the stream of values
+   */
+  virtual void dumpInternalVariables(std::map< std::string, std::string >& mapInternalVariables) = 0;
```
Comment (rosiereflo) [inline dynawo/sources/Modeler/Common/DYNModel.h:430]: No need to have this visible at that level. Specializing the dumpVariables of components of the model network should be enough

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/CPP/Common/DYNModelCPP.cpp [hunk 3]
```diff
@@ -56,7 +56,13 @@ ModelCPP::loadParameters(const string& /*parameters*/) {
 void
 ModelCPP::dumpVariables(map< string, string >& mapVariables) {
   stringstream values;
-  boost::archive::binary_oarchive os(values);
+  dumpVariablesInStream(values);
+  mapVariables[ variablesFileName() ] = values.str();
+}
+
+void
+ModelCPP::dumpVariablesInStream(stringstream& streamVariables) {
```
Comment (rosiereflo) [inline dynawo/sources/Models/CPP/Common/DYNModelCPP.cpp:64]: const?

File: nrt/data/DynaWaltz/Nordic_DumpLoad/Events.dyd [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2023, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline nrt/data/DynaWaltz/Nordic_DumpLoad/Events.dyd:3]: 2024

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Models/CPP/Common/DYNModelCPP.h [hunk 5]
```diff
@@ -84,16 +84,12 @@ class ModelCPP : public SubModel {
   double evalCalculatedVarI(unsigned iCalculatedVar) const override = 0;
 
   /**
-   * @brief export the parameters of the sub model for dump
-   *
-   * @param mapParameters : map associating the file where parameters should be dumped with the stream of parameters
+   * @copydoc SubModel::dumpParameters(std::map< std::string, std::string >& mapParameters)
```
[the remark this answers is not in the document]
Comment (vermeulenthi) [reply dynawo/sources/Models/CPP/Common/DYNModelCPP.h]: ok, changed back.

File: dynawo/sources/Models/CPP/Common/DYNModelCPP.cpp (hunk 3 above)
Comment (vermeulenthi) [reply dynawo/sources/Models/CPP/Common/DYNModelCPP.cpp]: const added

File: nrt/data/DynaWaltz/Nordic_DumpLoad/Events.dyd (hunk 4 above)
Comment (vermeulenthi) [reply nrt/data/DynaWaltz/Nordic_DumpLoad/Events.dyd]: 2024 added

**Review by rosiereflo — APPROVED** [review]
Looks good. Did you run the internal tests?

**Comment by vermeulenthi** [comment]
@rosiereflo Yes, tests and NRT ok.

**Pull request merged**